Upstream nvim-treesitter utils to Neovim

qrasmont opened the pull request:
Closes #17974


The following is a list of all the utility functions found in nvim-treesitter. As mentioned in the issue, this does not mean all should be upstreamed as is or at all. Some change to `languagetree` could be necessary to simplify/optimize some of them.

This is a starting point to do some triage as I dive into each function and determine what and how to add each one (or with any of your suggestions)

## From [nvim-treesitter/lua/nvim-treesitter/ts_utils.lua](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/ts_utils.lua):

### Will not upstream
- get_next_node() -> node:next_named_sibling() exists + opts not useful enough
- get_previous_node() -> node:prev_named_sibling() exists + opts not useful enough
- get_node_text() (function with same name but different behavior exists, see [comment below](https://github.com/neovim/neovim/pull/18232#issuecomment-1107800511), option has been added to allow to get nvim-treesitter behavior)
- get_root_for_position() -> not needed with languagetree changes
- is_in_node_range() -> not needed with languagetree changes
- memoize_by_buf_tick()


### As a node methods
- [x] get_named_children() -> `node:named_children()` (could imply  `node:children()` [comment below](https://github.com/neovim/neovim/pull/18232#issuecomment-1110102914))
- [x] get_root_for_node() -> `node:root()`
- [x] node_length() -> `node:length`

### Will upstream
- [x] is_parent()
- [x] get_node_at_cursor()
- [x] highlight_range()
- [x] highlight_node()
- [x] get_vim_range()
- [x] update_selection()
- [x] goto_node()
- [x] node_to_lsp_range()
- [x] swap_nodes()
- [x] get_node_range()

## From [playground//lua/nvim-treesitter-playground/utils.lua](https://github.com/nvim-treesitter/playground/blob/master/lua/nvim-treesitter-playground/utils.lua):
### Will upstream
- [x] get_hl_groups_at_position()
- [x] node_contains()

### TODO
- [x] Fix coding style
- [x] https://github.com/neovim/neovim/pull/18232#issuecomment-1143252718
- [ ] ~~Tests~~

theHamsta commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -114,4 +114,24 @@ function M.get_string_parser(str, lang, opts)
+--- Determines whether a node is the parent of another
+-- @param dest the possible parent
+-- @param source the possible child node
+function M.is_parent(dest, source)
```
Probably `is_ancestor` would be more clear

clason replied:
I think "parent" is the standard terminology for trees in CS?

clason commented:
Note that we probably don't want to upstream _all_ util functions indiscriminately. It would be best to do one function per commit (including other necessary functions to make each commit work) so they can be discussed individually. (That was likely your plan, anyway -- just wanted to make sure.)

qrasmont commented:
> Note that we probably don't want to upstream _all_ util functions indiscriminately. It would be best to do one function per commit (including other necessary functions to make each commit work) so they can be discussed individually. (That was likely your plan, anyway -- just wanted to make sure.)

Yes that's the plan, I just quickly opened the PR in draft to notify I started working on it.

gpanders commented:
Thanks for taking this on! Looking forward to it.

qrasmont commented:
> Thanks for taking this on! Looking forward to it.

My pleasure :wink: 

Should I also start adding some tests in the process ?
I see there is only one function tested in `ts_utils` (it might not make sense for all of them of course)

clason commented:
Oh, you walked right into this one...

Yes please -- if it can be tested, it should be tested!

clason commented:
About `get_node_text()`: this already has an upstream version under `vim.treesitter.query`, which unfortunately behaves different from the (older) nvim-treesitter one: the former returns a concatenated string, while the latter returns a table of strings. (This lead to much breakage when trying to replace the latter with the former...)

I'd propose extending the function here to accept an option table with  `{ concat = true }` (by default -- name can be bikeshedded) that allows skipping the concat https://github.com/neovim/neovim/blob/843d9223a14c599fda1b2abf1e7072be738fb831/runtime/lua/vim/treesitter/query.lua#L211 to avoid having to re-split it downstream. This would allow unifying both functions and finally replacing the downstream one; see https://github.com/nvim-treesitter/nvim-treesitter/pull/2838#issuecomment-1102830511 for context.

(I also feel that this doesn't really fit into this module; I'd move it to wherever we decide the upstreamed modules should live and, if necessary, just leave a compat stub under `query`.)

clason commented on runtime/doc/treesitter.txt:
```diff
@@ -378,6 +378,15 @@ get_string_parser({str}, {lang}, {opts})                 *get_string_parser()*
+
+                Parameters: ~
+                    {dest}    The destination node (the potential ancestor)
+                    {source}  The source node (the potential child)
```
```suggestion
                    {source}  The source node (the potential descendant)
```

clason replied:
(will be overwritten by the doc generation script anyway)

qrasmont replied:
Oh Ok I just found out about `gen_vimdoc.py` ^^ In that case I guess I'll make a final commit with its output for `runtime/doc/treesitter.txt`

clason commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -114,4 +114,24 @@ function M.get_string_parser(str, lang, opts)
 
+--- Determines whether a node is the ancestor of another
+-- @param dest the possible ancestor
+-- @param source the possible child node
```
```suggestion
-- @param source the possible descendant node
```

qrasmont commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -137,4 +137,45 @@ function M.is_ancestor(dest, source)
+---                  to return the next parent as the next node if it has no children.
+---
+---@returns The next node
+function M.get_next_node(node, opts)
```
I used an opts table here as it seems like something I see a lot in other APIs when a function has optional parameters, good or bad idea ?

clason replied:
Good, in general. It's most used for API functions with stability guarantees, but widely preferred for, well, optional arguments.

(Rule of thumb: mandatory is positional, optional keyword style -- which in Lua means part of an opts table)

theHamsta commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -114,4 +114,106 @@ function M.get_string_parser(str, lang, opts)
+---                  - switch_parent: (boolean, default false) Allow to return the last node from
+---                  the previous parent if the start node is the first node of its parent.
+---                  - return_prev_parent: (boolean, default false) With switch_parent = true, allow
+---                  to return the previous parent as the previous node if it has no children.
```
I think those options are not really necessary. Those are just a relic of https://github.com/theHamsta/crazy-node-movement wouldn't know whether this is useful enough for core

bfredl replied:
without options this should be the same as builtin `node:prev_[named_]sibling()`

qrasmont replied:
In that case, if the options are not interesting to upstream, I would drop it entirely. Same for `get_next_node`

theHamsta replied:
I think prev_sibling didn't exist at the time this was added to nvim-treesitter

bfredl commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -114,4 +114,106 @@ function M.get_string_parser(str, lang, opts)
+---                  to return the next parent as the next node if it has no children.
+---
+---@returns The next node
+function M.get_next_node(node, opts)
```
is this the same as `node:next_sibling()` when opts are not set? could be mentioned in the docs, as it is more efficient then.

clason replied:
...or consolidated?

clason commented:
Should probably also standardize on method vs. function syntax, while we're at it?

qrasmont commented:
> Should probably also standardize on method vs. function syntax, while we're at it?

Do you mean for example, deciding to integrate a utility function as a node method instead of a function that takes a node as a parameter ?

qrasmont commented:
> > Should probably also standardize on method vs. function syntax, while we're at it?
> 
> Do you mean for example, deciding to integrate a utility function as a node method instead of a function that takes a node as a parameter ?

With this in mind, regarding: [get_named_children](https://github.com/nvim-treesitter/nvim-treesitter/blob/82e43748cf70adab261ba0a23a342774bd5b9088/lua/nvim-treesitter/ts_utils.lua#L146-L152)
```lua
function M.get_named_children(node)
  local nodes = {}
  for i = 0, node:named_child_count() - 1, 1 do
    nodes[i + 1] = node:named_child(i)
  end
  return nodes
end
```
It could become `node:named_children()` (in which case, you would expect to have `node:children()` as well)

Would that be preferable ?

qrasmont committed:
feat(treesitter): upstream get_node_at_cursor()

Util from the nvim-treesitter project.

qrasmont commented:
I'm pushing those last 2 commits, while I'm still working on them to get some feedback on the direction I'm taking.
The idea being to extend languagetree to have a more direct way to get useful data (based on a given range)
In this case allowing to simplify the logic inside `get_node_at_cursor`

qrasmont committed:
feat(treesitter): upstream get_vim_range()

Util from the nvim-treesitter project.

qrasmont committed:
feat(treesitter): upstream highlight_range()

Util from the nvim-treesitter project.

clason commented:
Regarding `get_node_range`, see #18276

qrasmont committed:
feat(treesitter): upstream highlight_node()

Util from the nvim-treesitter project.

qrasmont committed:
fixup! feat(treesitter): upstream update_selection()

Util from the nvim-treesitter project.

qrasmont committed:
feat(treesitter): upstream update_selection()

Util from the nvim-treesitter project.

qrasmont committed:
feat(treesitter): upstream goto_node()

Util from the nvim-treesitter project.

clason commented:
> `get_node_range()` -> added private with https://github.com/neovim/neovim/pull/18276 (TBD make public)

Is that even doing the same thing? It shouldn't preclude upstreaming/consolidating the nvim-treesitter function. (May have to be renamed to avoid collision.)

clason commented:
(And would be great if you could add EmmyLua-style type annotations to the docstrings while you're at it; we're incrementally fixing the documentation in this regard.)

qrasmont committed:
feat(treesitter): upstream node_to_lsp_range()

Util from the nvim-treesitter project.

qrasmont committed:
feat(treesitter): upstream swap_nodes()

Util from the nvim-treesitter project.

clason commented:
@qrasmont it looks like the last commit broke CI -- could you bisect to make sure? (From the output, it looks like it's actually broken code rather than changed behavior.)

qrasmont commented:
> @qrasmont it looks like the last commit broke CI -- could you bisect to make sure? (From the output, it looks like it's actually broken code rather than changed behavior.)

Yes I indeed broke the functional test with the last commit.

clason commented:
Without looking to deeply at the code, I suspect some sort of circular inclusion -- can you try replacing the `highlighter.active` with `M.highlighter.active` instead of requiring `vim.treesitter.highlighter` at the top?

stasjok commented:
Maybe replace `get_node_at_cursor` with a more general `get_node_at_pos` which defaults to current cursor position? For example I had a trouble with `get_node_at_cursor` for some language (don't remember which exactly) in insert mode, because a current node ends before a cursor (in insert node current position is next character, e.g. if line has 3 chars, then current pos in insert mode is 3, but a tree-sitter node ended at 2). So in that time I really wanted to provide position minus one to `get_node_at_cursor`.

qrasmont commented:
> @qrasmont what are your plans with this PR? Will you be able to finish it? (There are some other changes in the works that are waiting on this upstreaming.)
> 
> Maybe we could finish up the current state (sans the failing commit) and leave the remaining functions for a follow-up PR?

Sorry, It's been hard to make time to work on it these past few weeks. I should be able to resume work on it mid next week and have it ready for review by the end of next weekend. It's essentially lacking some additional tests and resolving the `get_node_range` dual implementation.

clason commented:
No worries, this is completely understandable! Let me know when I should look at something.

qrasmont commented:
@clason I need to double check a potential issue with `get_hl_groups_at_position` (always getting an empty table in my tests) but otherwise ready for review.

(I ended up not adding any new tests but will do if requested)

qrasmont commented:
I have not yet figured out what's wrong with `get_hl_groups_at_position` but using the original function from [nvim-treesitter-playground](https://github.com/nvim-treesitter/playground/blob/master/lua/nvim-treesitter-playground/utils.lua) on master behaves similarly.
Still looking.

AckslD commented:
@qrasmont I opened https://github.com/nvim-treesitter/nvim-treesitter/pull/3233 which has a small improvement to `update_selection`. @theHamsta mentioned that it could be worth to suggest this change also here. If this change makes sense to everyone, should I then open a MR to this branch or would you incorporate it?

qrasmont commented:
> @qrasmont I opened [nvim-treesitter/nvim-treesitter#3233](https://github.com/nvim-treesitter/nvim-treesitter/pull/3233) which has a small improvement to `update_selection`. @theHamsta mentioned that it could be worth to suggest this change also here. If this change makes sense to everyone, should I then open a MR to this branch or would you incorporate it?

It's small enough, I'll add it in :+1:

clason commented:
bump -- what is the current state? can we get this PR in a mergable state if some things are left for followup PRs?

qrasmont committed:
fix(docs): update auto-generated documentation

qrasmont commented:
- I fixed the conflict with what's been merged in https://github.com/neovim/neovim/pull/19946
- The change suggested by @AckslD has been added to update_selection()
- I dropped get_hl_groups_at_position while I investigate what's wrong and will submit it in a separate PR
- API docs have been updated

clason commented:
Thank you!

Yes, get_hl_at_position is no longer needed since the PR that enabled mapping captures directly.

A new convenience function is still possible but can be added later.

clason commented on test/functional/treesitter/utils_spec.lua:
```diff
@@ -30,4 +30,20 @@ describe('treesitter utils', function()
   end)
+
+  it('can convert a ts range to a vim range', function()
+    insert([[
```
This needs to be gated behind a 
```lua
    if pending_c_parser(pending) then return end
```
since the `functionaltest-lua` action doesn't have a parser installed (for now).

clason commented:
Otherwise LGTM @neovim/treesitter

clason committed:
fixup! feat(treesitter): upstream get_vim_range()

clason commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -241,4 +241,223 @@ function M.get_captures_at_position(bufnr, row, col)
+---@param hlgroup string The highlight group name
+function M.highlight_range(bufnr, range, ns, hlgroup)
+  local start_row, start_col, end_row, end_col = unpack(range)
+  vim.highlight.range(bufnr, ns, hlgroup, { start_row, start_col }, { end_row, end_col })
```
```suggestion
  vim.highlight.range(bufnr, ns, hlgroup, { start_row, start_col }, { end_row, end_col }, { priority = vim.highlight.treesitter })
```

theHamsta replied:
The idea was actually to use this function in user code and thus have the normal higher user priority to paint over tree-sitter. Basically to import a tree-sitter range for non-tree-sitter highlights. Priority could go into an opts argument

clason replied:
Makes sense. Let's add the priority `opts` then and document that it defaults to `user`.

clason committed:
fixup! feat(treesitter): upstream highlight_range()

clason commented:
@qrasmont After some discussion, I think we should start with upstreaming only those utilities relevant to highlighting (which is included in core). This means I will rebase and squash this PR omitting 
* `update_selection` (only used for incremental selection)
* `swap_nodes` (only used for textobjects)
* `goto_node` (only used for playground)
* `node_to_lsp_range` (only used for `swap_nodes`)
* `vim_range` (only used for `goto_node` and `update_selection`)

I apologize for not checking these functions earlier! You might want to keep the commits around in case we want to upstream them later.

AckslD commented:
> update_selection (only used for incremental selection)

Maybe irrelevant for what will be included and maybe already clear but note that `update_selection` is also used by [`nvim-treesitter-textobjects`](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/lua/nvim-treesitter/textobjects/select.lua#L101)

clason commented:
Change of plan: For 0.8, we will just include `get_node_at_position` (slight refactor, which I'll break out into a separate PR) and leave the rest for post-0.8 (when we may add more treesitter functionality).

justinmk commented on runtime/doc/treesitter.txt:
```diff
@@ -382,6 +382,19 @@ get_captures_at_position({bufnr}, {row}, {col})
         (table) A table of captures
 
+get_node_at_cursor({winnr}, {opts})                     *get_node_at_cursor()*
+    Gets the smallest named node under the cursor
```
why does this depend on the cursor instead of just taking a position arg?

clason replied:
Yes, that is going to be refactored, see my last comment.

justinmk commented on runtime/doc/treesitter.txt:
```diff
@@ -441,6 +499,40 @@ node_contains({node}, {range})                               *node_contains()*
+
+    Parameters: ~
+        {bufnr}                (number) The buffer number
+        {node}                 (table) The node to visually select select
```
typo

justinmk commented on runtime/doc/treesitter.txt:
```diff
@@ -441,6 +499,40 @@ node_contains({node}, {range})                               *node_contains()*
+                                 second node after the swap
+
+update_selection({bufnr}, {node}, {opts})                 *update_selection()*
+    Sets visual selection for node
```
this needs a bit more explanation. Is it setting visual selection based on the parsed scope represented by a TS node? Then we don't need this. We need : 

1. a function that sets visual selection from a "region" (related: https://github.com/neovim/neovim/pull/13896 )
2. a function that gets a "region" from a TS node

clason replied:
Yes, that is a function we're probably not going to upstream (in this form at least).

(2. is already available as `get_node_range`)

AckslD commented:
@clason @qrasmont I opened https://github.com/nvim-treesitter/nvim-treesitter/pull/3487 to make it easier to work with ranges of query matches (in particular if `#offset!` is used). It adds some functions that uses `metadata.range` if it exists and otherwise the range of the `node`. It would then also be nice if functions here such as `get_node_range` either could be refactored to
1. take a node or a range
2. or extract the main function to a function that takes a range, eg `get_range_text`.

Any thoughts on this?

justinmk commented:
Can't we close this? What else is left to upstream (that is actually wanted)?
